circuit-control.ino

```diff
@@ -162,248 +83,40 @@ void setup() {
-        }
-
-        // if user turned on inspiratory hold, transition to INSP_HOLD_STATE
-        if (display.inspHold()) { 
-          setState(HOLD_INSP_STATE); 
-          beginHoldInspiration(); 
-        } else {
-          // otherwise, transition to EXP_STATE
-          setState(EXP_STATE);
-          beginExpiration();
-        }
-
-        inspDuration = millis() - cycleTimer; // Record length of inspiration
-      } else {
-        // if transition criteria is not met, keep adjusting inspiratory valve
-        inspValve.maintainBreath(cycleTimer);
-      }
-      break;
-
-    case HOLD_INSP_STATE:
-      display.updateFlowWave(inspFlowReader.get()); 
-
-      // if we reached the end time for HOLD_INSP_STATE
-      if (HOLD_INSP_DURATION <= millis() - inspHoldTimer) {
-        inspPressureReader.setPlateau();
-        // check plateau pressure range and alarm if abnormal  
-        if (inspPressureReader.plateau() > PPLAT_MAX) { 
-         alarmMgr.activateAlarm(ALARM_PPLAT_HIGH);
-        } else if (alarmMgr.alarmStatus(ALARM_PPLAT_HIGH)) {
-          alarmMgr.deactivateAlarm(ALARM_PPLAT_HIGH);
-        }
-
-        setState(EXP_STATE); // switch to EXP_STATE
-        beginExpiration();   
-      }
-      break;
-
-    case EXP_STATE:
-      // To update flow graph, flip sign of expiratory flow sensor to show flow out of lungs 
-      display.updateFlowWave(expFlowReader.get() * -1); 
-      expFlowReader.updateVolume();                     
-      
-      // if 80% of inspired volume has been expired, transition to PEEP_PAUSE_STATE 
-      if (expFlowReader.getVolume() >= targetExpVolume || millis() > targetExpEndTime + EXP_TIME_SENSITIVITY){ 
-        setState(PEEP_PAUSE_STATE); 
-        beginPeepPause();
-      }
-      break;
-
-    case PEEP_PAUSE_STATE:
-      // To update flow graph, flip sign of expiratory flow sensor to show flow out of lungs 
-      display.updateFlowWave(expFlowReader.get() * -1); 
-      expFlowReader.updateVolume();                     
-      
-      // if the PEEP pause time has run out, transition to HOLD_EXP_STATE
-      if (millis() - peepPauseTimer >= MIN_PEEP_PAUSE) {
-        expPressureReader.setPeep(); 
-        setState(HOLD_EXP_STATE);    
-      }
-      break;
-
-    case HOLD_EXP_STATE: {
-      display.updateFlowWave(expFlowReader.get() * -1); //update flow waveform on display
-      expFlowReader.updateVolume();                     //update expiratory flow counter
-
-      // Check if patient triggers inhale or state timed out 
-      bool patientTriggered = expPressureReader.get() < expPressureReader.peep() - display.sensitivity();
-      bool timeout = (millis()  > targetCycleEndTime); 
-
-      if (patientTriggered || timeout) { 
-        beginInspiration();   
-        setState(INSP_STATE);
-      }
-
-    } break;
-  } // End switch
+  // handle state maintenance & update
+  state = state->update();
```

Comment:
Instead of having each state call `enter` for the next state, consider doing it here, in the main loop or, better yet, in a subroutine:
```
void runStateMachine() {
  State* nextState = state->update();
  while (nextState != state) {
    state = nextState;
    nextState = state->enter();
  }
}
```
Note that this neatly solves the issue of whether `enter` should be virtual.  Inside the individual states, instead of writing:
```
    nextState = InsHoldStateVC::enter();
    ...
    return nextState;
```
You would write simply:
```
    return &InsHoldStateVC::singleton();
```
I like centralizing the state machine logic this way, and I think it will be easier to maintain. Also, you can eventually make the state machine itself into a class that holds on to the current state.

Reply:
As you can see, I adopted a lot of these ideas in a recent commit.